pkg/networkserver/grpc_deviceregistry.go

```diff
@@ -45,82 +46,206 @@ func (ns *NetworkServer) Set(ctx context.Context, req *ttnpb.SetEndDeviceRequest
+			return nil, nil, errInvalidFieldValue.WithAttributes("field", "mac_settings.desired_rx2_frequency")
+		}
+
+		sets := append(req.FieldMask.Paths[:0:0], req.FieldMask.Paths...)
+		if ttnpb.HasAnyField(sets, "version_ids") {
 			// TODO: Apply version IDs (https://github.com/TheThingsIndustries/lorawan-stack/issues/1544)
 		}
 
-		if err := ttnpb.RequireFields(paths,
+		if err := ttnpb.RequireFields(sets,
 			"frequency_plan_id",
 			"lorawan_phy_version",
 			"lorawan_version",
-			"mac_settings.use_adr",
-			"resets_f_cnt",
-			"resets_join_nonces",
-			"supports_class_b",
-			"supports_class_c",
 			"supports_join",
-			"uses_32_bit_f_cnt",
 		); err != nil {
 			return nil, nil, errInvalidFieldMask.WithCause(err)
 		}
-
-		if req.Device.MACSettings == nil {
-			return nil, nil, errNoMACSettings
+		if len(req.Device.FrequencyPlanID) == 0 {
+			return nil, nil, errInvalidFieldValue.WithAttributes("field", "frequency_plan_id")
 		}
-
-		if !ttnpb.HasAnyField(paths, "mac_settings.adr_margin") {
-			// TODO: Apply NS-wide default (https://github.com/TheThingsIndustries/lorawan-stack/issues/1544)
-			req.Device.MACSettings.ADRMargin = 15
-			paths = append(paths, "mac_settings.adr_margin")
-		} else if req.Device.MACSettings.ADRMargin == 0 {
-			return nil, nil, errInvalidADRMargin
+		if err := req.Device.LoRaWANVersion.Validate(); err != nil {
+			return nil, nil, errInvalidFieldValue.WithAttributes("field", "lorawan_version").WithCause(err)
 		}
-
-		if !ttnpb.HasAnyField(paths, "mac_settings.class_b_timeout") {
-			// TODO: Apply NS-wide default if not set (https://github.com/TheThingsIndustries/lorawan-stack/issues/1544)
-			req.Device.MACSettings.ClassBTimeout = time.Minute
-			paths = append(paths, "mac_settings.class_b_timeout")
-		} else if req.Device.MACSettings.ClassBTimeout == 0 {
-			return nil, nil, errInvalidClassBTimeout
+		if req.Device.LoRaWANPHYVersion == ttnpb.PHY_UNKNOWN {
+			return nil, nil, errInvalidFieldValue.WithAttributes("field", "lorawan_phy_version")
 		}
 
-		if !ttnpb.HasAnyField(paths, "mac_settings.class_c_timeout") {
-			// TODO: Apply NS-wide default if not set (https://github.com/TheThingsIndustries/lorawan-stack/issues/1544)
-			req.Device.MACSettings.ClassCTimeout = 10 * time.Second
-			paths = append(paths, "mac_settings.class_c_timeout")
-		} else if req.Device.MACSettings.ClassCTimeout == 0 {
-			return nil, nil, errInvalidClassCTimeout
+		if ttnpb.HasAnyField(sets, "supports_class_b") && req.Device.SupportsClassB {
+			if err := ttnpb.RequireFields(sets,
+				"mac_settings.ping_slot_date_rate_index",
+				"mac_settings.ping_slot_frequency",
+				"mac_settings.ping_slot_periodicity",
+			); err != nil {
+				return nil, nil, errInvalidFieldMask.WithCause(err)
+			}
 		}
 
 		if req.Device.SupportsJoin {
-			return &req.Device, paths, nil
+			var hasSession bool
+			for _, p := range sets {
+				if p == "session" || strings.HasPrefix(p, "session.") {
```

Comment:
Don't we have a helper for this?

Reply:
Unfortunately not, I tried using `ttnpb.HasAnyField(sets, "session")`, but that caused bugs (https://github.com/TheThingsNetwork/lorawan-stack/issues/104#issuecomment-465074076)
We need to check if `sets` contains any or all of `session` fields, this was the easiest way to do this and it looked like this feature is not required by other components, so I did not add a helper for it. I was also confused about how to name the function, since we already have `HasAnyField`.

Reply:
You can invert it; `ttnpb.HasAnyField([]string{"session"}, sets...)`